_posts/2018-04-09-middleware-event.md

```diff
@@ -0,0 +1,191 @@
+the [documentation](http://docs.simplebus.io/en/latest/index.html). 
+Maybe you want to log something specific in your system, or maybe you are building a gamification application that gives
+users experience points based on their actions. Then using SimpleBus, together with your events and commands might be just
+right for you. 
+
+## How it works
+The main point is to do some kind event manipulation on those events that have already been dispatched inside your application. 
+Maybe you want to save the events to a log, or maybe add some more data to the events before passing them on to another 
+application.
+Below is a section of the middleware class, the entire middleware is shown at the bottom of this post. 
+
+{% highlight php %}
+$eventData = [
+    // .. Add some event data
+]
+// Add the data to a DataContainer object
+$container = new DataContainer($eventData);
+
+$this->eventDispatcher->dispatch('foo', new Foo($event, $container));
+{% endhighlight %}
+
+### The dispatched event
+The Foo event class uses the DataContainer object instead of a
+simple array, since an array is passed by value and not by reference. Therefore by using the DataContainer object we can 
+make sure that the passed parameter has the correct value, since objects are passed by reference, which means that if the 
+variable gets modified, the value will be updated for both parties using it. 
+
+{% highlight php %}
+
+use Symfony\Component\EventDispatcher\Event;
+
+class Foo extends Event
+{
+    private $event;
+    private $container;
+
+    public function __construct($event, DataContainer $container)
+    {
+        $this->event = $event;
+        $this->container = $container;
+    }
+
+    public function getEvent()
+    {
+        return $this->event;
+    }
+
+    public function getContainer(): DataContainer
+    {
+        return $this->container;
+    }
+}
+{% endhighlight %}
+
+### The DataContainer class
+Below is the DataContainer class, used to add some new data to the existing event. 
+
+{% highlight php %}
+
+class DataContainer
+{
+    /** @var array */
+    private $data;
+
+    public function __construct(array $data)
+    {
+        $this->data = $data;
+    }
+
+    public function addData(string $key, $value)
+    {
+        $this->data[$key] = $value;
+    }
+}
+
+{% endhighlight %}
+
+## Manipulate the event
+Now you will need something that can manipulate the existing event. Creating an event subscriber that subscribes to the 'foo'
+event is a good start. Now it's up to you to decide what your subscriber should do.     
+
+In this case imagine that we need a users age and address for later use, but they are both missing in the SimpleBus event data that
+was provided from our application. This means that we somehow want to add these values to the event, before it gets passed
+on to another application. This will be done by the decorator class, which is a simple event subscriber. 
```

Comment:
You already said it is an "event subscriber". Or is this a different subscriber?